`flake.lib.nixosSystem`: Allow `nixpkgs.system` to be set modularly; improve error message

roberth opened the pull request:
###### Description of changes

Got a question today from someone who wanted to set `nixpkgs.system` modularly in a flake. They thought the `system` argument was the only way to set it, which it is not.

This is change is in line with the improvements that were intended for `eval-config.nix` but weren't implemented, presumably because of compatibility concerns.

Note that this is a simplification of the `nixosSystem` _interface_, relying on established options rather than repeating parameters for no good reason. The implementation could eventually be simplified too, by extracting a new entrypoint from `eval-config.nix`, similar to how `eval-config-minimal.nix` was extracted. I'm not doing that here because that would be a distraction.

The new error message when `system` is forgotten is:

```
nix-repl> (lib.nixosSystem { modules = [ { } ]; }).config.system.build.toplevel
error: The option nixpkgs.system has not been set. Please set it, or set nixpkgs.hostPlatform instead. Note that nixos-generate-config will set the latter for you in hardware-configuration.nix.
```

the old message was

```
error: attribute 'currentSystem' missing
```



###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
nixos/eval-config: Allow system to be set modularly when arg is null

roberth committed:
flake.lib.nixosSystem: Allow nixpkgs.system to be set instead

roberth committed:
flake.nix: Format

roberth committed:
nixos/nixpkgs.nix: Recommend hostPlatform instead of system

The ${opt.*} syntax will print the full path when NixOS is used
as a submodule.

nixpkgs.system / nixpkgs.localSystem must not be read by any
other module because its meaning is ambiguous in cross vs
non-cross contexts. hostPlatform is generally what you need.
*Where* you build something generally doesn't matter in a
system _configuration_ context like NixOS.

infinisil commented on nixos/modules/misc/nixpkgs.nix:
```diff
@@ -280,6 +280,11 @@ in
       example = "i686-linux";
+      default =
+        throwIf
+          (opt.system.highestPrio == (lib.mkOptionDefault {}).priority)
```
This is rather ugly. It's cleaner to change the type of `system` to `nullOr str` and having a default of `null` and using `assertions`

roberth replied:
Hmm. The really ugly bit is that `default = throw "..."` did not work as one might expect.
Otherwise I thought that this localized solution was quite on point.

roberth replied:
It was an issue with the derivation scrubbing logic in the docs. I've simplified this.

roberth commented on nixos/modules/misc/nixpkgs.nix:
```diff
@@ -280,6 +280,11 @@ in
+      default =
+        throwIf
+          (opt.system.highestPrio == (lib.mkOptionDefault {}).priority)
+          ''The option ${opt.system} has not been set. Please set ${opt.hostPlatform} instead. Note that nixos-generate-config will set it in hardware-configuration.nix.''
```
Setting `system` is still valid for now, as we can't start an ecosystem-wide change without first making the new way widely available. Even adding a warning will cause annoyance as 2nd-party repos can not implement the change without breaking 22.05 support. So this will take time. We can consider this for 23.05, and unstable in January when 22.05 is EOL.

I've put in place migration logic for such changes, but I didn't use it here because the implementation is complicated enough as it is. See https://github.com/NixOS/release-wiki/issues/31.
We could add such a warning in advance with `isInOldestRelease`, but I'm hesitant to automate this before this has ecosystem support.

infinisil replied:
I'm mainly asking because of the sentence itself, which doesn't make sense to me. If it's complaining about `system` _not_ being set, then the natural solution would be to _set_ `system`. The sentence only makes sense to me if you remove the first "not".

roberth replied:
@infinisil Could you review again?

I've taken into account your feedback that it's ok to set `system`.
Setting `hostPlatform` is the default solution for an unset `system` in the case of `nixos-generate-config`-based configurations, so this should be mentioned. Eventually we can deprecate `system` and get rid of this complexity.

The new message is:

`The option nixpkgs.system has not been set. Please set it, or set nixpkgs.hostPlatform instead. Note that recent versions of nixos-generate-config will set the latter for you in hardware-configuration.nix.`

alyssais replied:
Maybe we could explain when somebody would want to use one or the other?  In particular, push people towards setting `hostPlatform` if they don't need to support 22.05?  Otherwise, people will see this message, set `nixpkgs.system` because it's the first thing mentioned, then have to change it _again_ when we deprecate `nixpkgs.system`.  We could save them that trouble by encouraging them to use `hostPlatform` if they don't need 22.05 support.

roberth replied:
Yeah I was trying to be too conservative (and too concise for that matter). The new message is:

```
error: Neither nixpkgs.hostPlatform nor or the legacy option nixpkgs.system has been set.
       You can set nixpkgs.hostPlatform in hardware-configuration.nix by re-running
       a recent version of nixos-generate-config.
       The option nixpkgs.system is still fully supported for NixOS 22.05 interoperability,
       but will be deprecated in the future, so we recommend to set nixpkgs.hostPlatform.
```

roberth commented on nixos/modules/misc/nixpkgs.nix:
```diff
@@ -280,7 +292,14 @@ in
       type = types.str;
       example = "i686-linux";
+      default = throw ''The option ${opt.system} has not been set. Please set ${opt.hostPlatform} instead. Note that nixos-generate-config will set it in hardware-configuration.nix.'';
+      defaultText = lib.literalMD ''
+        Traditionally `builtins.currentSystem`, but unset when invoking NixOS through `lib.nixosSystem`.
+      '';
       description = ''
+        This option does not need to be specified for NixOS configurations
+        with a recently generated <literal>hardware-configuration.nix</literal>.
+
```
![image](https://user-images.githubusercontent.com/496447/176198509-e4fda071-38e1-4400-b6a4-52694cc56b31.png)

alyssais reviewed (approved):
I'm trying to get us out of the business of passing system doubles around inside Nixpkgs, because not all platforms Nixpkgs supports can be expressed with a double.  It looks like this change only affects the entry point to NixOS, and will actually make it easier for a flake user to set e.g. `localSystem.config = "x86_64-unknown-linux-musl";` (this would also be important for all of the various ARM ABI variants), so this seems to be a step in the right direction.

Deprecating the `system` argument to eval-config.nix is definitely a good step, and something I was planning on proposing soon myself!

roberth committed:
nixos: Fix use of nixpkgs.localSystem

localSystem is ill-defined because unlike hostPlatform, its
meaning is different in a cross or non-cross context.

alyssais reviewed (approved):
Needs some squashing before being merged.

roberth commented:
> Needs some squashing before being merged.

Done. I find the remaining commit boundaries to be useful, as they form a build-up from low-level possibility to having the option to recommending it (subtly for the time being).

infinisil reviewed (approved):
Looking good!
